pkg/controllers/leaderworkerset_controller.go

```diff
@@ -523,8 +544,54 @@ func (r *LeaderWorkerSetReconciler) iterateReplicas(ctx context.Context, lws *le
 	return continuousReadyReplicas, lwsUnreadyReplicas, nil
 }
 
+func (r *LeaderWorkerSetReconciler) getLeaderStatefulSet(ctx context.Context, lws *leaderworkerset.LeaderWorkerSet) (*appsv1.StatefulSet, error) {
+	sts := &appsv1.StatefulSet{}
+	err := r.Get(ctx, types.NamespacedName{Name: lws.Name, Namespace: lws.Namespace}, sts)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	return sts, nil
+}
+
+func (r *LeaderWorkerSetReconciler) createControllerRevisionIfNonExist(ctx context.Context, sts *appsv1.StatefulSet, lws *leaderworkerset.LeaderWorkerSet) (*appsv1.ControllerRevision, error) {
+	if sts == nil {
+		return revisionutils.CreateLeaderWorkerSetRevision(ctx, r.Client, lws, "")
+	}
+
+	stsRevision, err := revisionutils.GetLeaderWorkerSetRevisionFromTemplateHash(ctx, r.Client, lws, sts.Labels[leaderworkerset.TemplateRevisionHashKey])
+	if err != nil {
+		return nil, err
+	}
+
+	if stsRevision == nil {
+		return revisionutils.CreateLeaderWorkerSetRevision(ctx, r.Client, lws, sts.Labels[leaderworkerset.TemplateRevisionHashKey])
+	}
+
+	return revisionutils.CreateLeaderWorkerSetRevision(ctx, r.Client, lws, "")
+}
+
+func (r *LeaderWorkerSetReconciler) leaderWorkerSetUpdated(ctx context.Context, sts *appsv1.StatefulSet, lws *leaderworkerset.LeaderWorkerSet, revision *appsv1.ControllerRevision) (bool, error) {
+	if sts == nil {
+		return false, nil
+	}
+
+	stsRevision, err := revisionutils.GetLeaderWorkerSetRevisionFromTemplateHash(ctx, r.Client, lws, sts.Labels[leaderworkerset.TemplateRevisionHashKey])
```

Comment:
The revision that is passed in is the one that matches the templateHash currently on the sts, so here we should be creating a NewRevision in-memory using the internal hash function and compare it with the one being passed in.

This function should also return this revision in case the comparison fails so that we create this updated revision in case it doesn't exist, and so instead of the boolean, we return a pointer to the updated revision if an update happened.

Reply:
So create a new revision, and call `EqualRevision()` on the new created one, and the sts one? I don't think that solves the bug. The bug that I see in the current implementation is that if the revision has a templateHash created before the CR change, once we create a new one to detect update, it will incorrectly detect it as it will now use the new hashing function. This can be fixed by removing the templateHash check from `equalRevision`. I don't see how this changes fixes that bug.